### ISSUE TITLE
"Untyped decorator makes function untyped" false positive on decorator with union type

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#16384]
```py
from typing import Callable, ParamSpec, TypeVar

T = TypeVar("T")
P = ParamSpec("P")

class C:
    def __call__(self, arg: Callable[P, T]) -> Callable[P, T]:
        return arg

foo: C | Callable[[Callable[P, T]], Callable[P, T]]

@foo
def bar() -> None:...
```
[playground](https://mypy-play.net/?mypy=master&python=3.11&flags=strict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Clocal-partial-types%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=958633e684515d96d8514e8090490e2d)

use case: i have to define a decorator conditionally that can be either one of these two very similar types
